neuralop/layers/spectral_convolution.py

```diff
@@ -451,20 +453,33 @@ def forward(
         
         weight = self.weight[slices_w]
 
+        ### Pick proper modes of fft
+
         # if separable conv, weight tensor only has one channel dim
         if self.separable:
             weight_start_idx = 1
         # otherwise drop first two dims (in_channels, out_channels)
         else:
             weight_start_idx = 2
-        starts = [(size - min(size, n_mode)) for (size, n_mode) in zip(list(x.shape[2:]), list(weight.shape[weight_start_idx:]))]
+        
         slices_x =  [slice(None), slice(None)] # Batch_size, channels
 
-        if self.complex_data:
-            slices_x += [slice(start//2, -start//2) if start else slice(start, None) for start in starts]
-        else:
-            slices_x += [slice(start//2, -start//2) if start else slice(start, None) for start in starts[:-1]]
-            slices_x += [slice(None, -starts[-1]) if starts[-1] else slice(None)] # The last mode already has redundant half removed
+        for all_modes, kept_modes in zip(fft_size, list(weight.shape[weight_start_idx:])):
+            # 0th frequency is located at n // 2 in each direction
+            # we grab n//2 - n_modes // 2, n//2 + n_modes // 2 for each dim
+            # if we keep an odd number of modes, grab one extra positive frequency
+            center = all_modes // 2
+            positive_freqs = negative_freqs = kept_modes // 2
+            if kept_modes % 2 == 1:
+                positive_freqs += 1
+            slices_x += [slice(center - negative_freqs, center + positive_freqs)]
+        
+        if not self.complex_data:
+            if weight.shape[-1] < fft_size[-1]:
```

Comment:
Don't we potentially need this check with respect to weight shape for all cases, including complex?

Reply:
You're right, thanks for the catch